icicle/include/hash/hash.cuh

```diff
@@ -0,0 +1,257 @@
+  default_sponge_config(const device_context::DeviceContext& ctx = device_context::get_default_device_context())
+  {
+    SpongeConfig config = {
+      ctx,   // ctx
+      false, // are_inputs_on_device
+      false, // are_outputs_on_device
+      0,     // input_rate
+      0,     // output_rate
+      0,     // offset
+      false, // recursive_squeeze
+      false, // recursive_squeeze
+      false, // is_async
+    };
+    return config;
+  }
+
+  template <typename H, typename PreImage, typename Image>
+  class SpongeHasher
+  {
+  public:
+    unsigned int get_width() const {
+      return static_cast<const H*>(this)->width;
+    }
+
+    unsigned int get_preimage_max_length() const {
+      return static_cast<const H*>(this)->preimage_max_length;
+    }
+
+    cudaError_t allocate_states(Image** states, unsigned int number_of_states, const device_context::DeviceContext& ctx) const
+    {
+      CHK_INIT_IF_RETURN();
+      return cudaMallocAsync(states, number_of_states * get_width() * sizeof(Image), ctx.stream);
+    }
+
+    cudaError_t free_states(Image* states, const device_context::DeviceContext& ctx) const
+    {
+      CHK_INIT_IF_RETURN();
+      return cudaFreeAsync(states, ctx.stream);
+    }
+
+    cudaError_t absorb_many(
+      const PreImage* inputs,
+      Image* states,
+      unsigned int number_of_states,
+      unsigned int input_block_len,
+      const SpongeConfig& cfg) const
+    {
+      CHK_INIT_IF_RETURN();
+      unsigned int width = get_width();
+      unsigned int preimage_max_length = get_preimage_max_length();
+
+      if (cfg.input_rate * sizeof(PreImage) > preimage_max_length * sizeof(Image))
+        THROW_ICICLE_ERR(IcicleError_t::InvalidArgument, "Input rate can not be bigger than preimage max length");
+      if (input_block_len > cfg.input_rate)
+        THROW_ICICLE_ERR(IcicleError_t::InvalidArgument, "Input blocks can not be bigger than hash rate");
+
+      // This allows to copy hash inputs and apply zero padding
+      CHK_IF_RETURN(cudaMemcpy2DAsync(
+        states, width * sizeof(Image),             // (Dst) States pointer and pitch
+        inputs, cfg.input_rate * sizeof(PreImage), // (Src) Inputs pointer and pitch
+        input_block_len * sizeof(PreImage),        // Width of the source matrix
+        number_of_states,                          // Height of the source matrix
+        cfg.are_inputs_on_device ? cudaMemcpyDeviceToDevice : cudaMemcpyHostToDevice, cfg.ctx.stream));
+
+      cudaError_t padding_error =
+        static_cast<const H*>(this)->pad_many(states, number_of_states, input_block_len, cfg.input_rate, cfg.ctx);
+      CHK_IF_RETURN(padding_error);
+
+      cudaError_t permutation_error =
+        static_cast<const H*>(this)->run_permutation_kernel(states, states, number_of_states, cfg.aligned, cfg.ctx);
+      CHK_IF_RETURN(permutation_error);
+
+      if (!cfg.is_async) CHK_IF_RETURN(cudaStreamSynchronize(cfg.ctx.stream));
+
+      return CHK_LAST();
+    }
```

Comment:
it is one absorb stage right?

Reply:
yes, you can run it multiple times to absorb more inputs